src/rpc/RPCHelpers.cpp

```diff
@@ -405,61 +407,111 @@ getStartHint(ripple::SLE const& sle, ripple::AccountID const& accountID)
-        return Status{RippledError::rpcACT_NOT_FOUND};
+    auto const firstNFTPage = ripple::keylet::nftpage_min(accountID);
+    auto const lastNFTPage = ripple::keylet::nftpage_max(accountID);
 
-    auto const maybeCursor = parseAccountCursor(jsonCursor);
-    if (!maybeCursor)
-        return Status(ripple::rpcINVALID_PARAMS, "Malformed cursor");
+    // check if nextPage is valid
+    if (nextPage != beast::zero and firstNFTPage.key != (nextPage & ~ripple::nft::pageMask))
+        return Status{RippledError::rpcINVALID_PARAMS, "invalid Marker"};
 
-    auto [hexCursor, startHint] = *maybeCursor;
+    // no marker, start from the last page
+    ripple::uint256 currentPage = nextPage == beast::zero ? lastNFTPage.key : nextPage;
 
-    return traverseOwnedNodes(
-        backend,
-        ripple::keylet::ownerDir(accountID),
-        hexCursor,
-        startHint,
-        sequence,
-        limit,
-        jsonCursor,
-        yield,
-        atOwnedNode);
+    // read the current page
+    auto page = backend.fetchLedgerObject(currentPage, sequence, yield);
+
+    if (!page)
+    {
+        if (nextPage == beast::zero)  // no nft objects in lastNFTPage
+            return AccountCursor{beast::zero, 0};
+        else  // marker is in the right range, but still invalid
+            return Status{RippledError::rpcINVALID_PARAMS, "invalid Marker"};
+    }
+
+    // the object exists and the key is in right range, must be nft page
+    ripple::SLE pageSLE{ripple::SLE{ripple::SerialIter{page->data(), page->size()}, currentPage}};
+
+    auto count = 0;
+    // traverse the nft page linked list until the start of the list or reach the limit
+    while (true)
+    {
+        auto const nftPreviousPage = pageSLE.getFieldH256(ripple::sfPreviousPageMin);
+        atOwnedNode(std::move(pageSLE));
+        count++;
+
+        if (count == limit or nftPreviousPage == beast::zero)
+        {
+            if (nftPreviousPage != beast::zero)
+                return AccountCursor{nftPreviousPage, count};
+            else
+                return AccountCursor{beast::zero, count};
+        }
+
+        page = backend.fetchLedgerObject(nftPreviousPage, sequence, yield);
+        pageSLE = ripple::SLE{ripple::SerialIter{page->data(), page->size()}, nftPreviousPage};
+    }
+
+    return AccountCursor{beast::zero, 0};
 }
 
 std::variant<Status, AccountCursor>
-ngTraverseOwnedNodes(
+traverseOwnedNodes(
     BackendInterface const& backend,
     ripple::AccountID const& accountID,
     std::uint32_t sequence,
     std::uint32_t limit,
     std::optional<std::string> jsonCursor,
     boost::asio::yield_context& yield,
-    std::function<void(ripple::SLE&&)> atOwnedNode)
+    std::function<void(ripple::SLE&&)> atOwnedNode,
+    bool nftIncluded)
 {
     auto const maybeCursor = parseAccountCursor(jsonCursor);
+
+    if (!maybeCursor)
+        return Status{RippledError::rpcINVALID_PARAMS, "Malformed cursor"};
```

Comment:
I think other errors usually add a dot at the end of the error message. If that's the case we should probably try to keep it uniform. Unless of course rippled does it without the dot for this particular case.

Reply:
You are right. Shall be consistent. I will add dot